src/ekka_rlog_bootstrapper.erl

```diff
@@ -0,0 +1,212 @@
+%%
+%% Unless required by applicable law or agreed to in writing, software
+%% distributed under the License is distributed on an "AS IS" BASIS,
+%% WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+%% See the License for the specific language governing permissions and
+%% limitations under the License.
+%%--------------------------------------------------------------------
+
+%% @doc This module implements both bootstrap server and client
+
+-module(ekka_rlog_bootstrapper).
+
+%% API:
+-export([start_link/2, start_link_client/3]).
+
+%% gen_server callbacks:
+-export([ init/1
+        , terminate/2
+        , handle_call/3
+        , handle_cast/2
+        , handle_info/2
+        , code_change/3
+        ]).
+
+%% Internal exports:
+-export([do_push_batch/2, do_complete/3]).
+
+-include_lib("snabbkaffe/include/trace.hrl").
+
+%%================================================================================
+%% Type declarations
+%%================================================================================
+
+-type batch() :: { _From    :: pid()
+                 , _Table   :: ekka_rlog_lib:table()
+                 , _Records :: [tuple()]
+                 }.
+
+-record(server,
+        { shard       :: ekka_rlog:shard()
+        , subscriber  :: ekka_rlog_lib:subscriber()
+        , key_queue   :: replayq:q() | undefined
+        , tables      :: [ekka_rlog_lib:table()]
+        }).
+
+-record(client,
+        { shard       :: ekka_rlog:shard()
+        , server      :: pid()
+        , parent      :: pid()
+        }).
+
+%%================================================================================
+%% API funcions
+%%================================================================================
+
+%% @doc Start bootstrapper server
+-spec start_link(ekka_rlog:shard(), ekka_rlog_lib:subscriber()) -> {ok, pid()}.
+start_link(Shard, Subscriber) ->
+    gen_server:start_link(?MODULE, {server, Shard, Subscriber}, []).
+
+%% @doc Start bootstrapper client
+-spec start_link_client(ekka_rlog:shard(), node(), pid()) -> {ok, pid()}.
+start_link_client(Shard, RemoteNode, Parent) ->
+    gen_server:start_link(?MODULE, {client, Shard, RemoteNode, Parent}, []).
+
+%%================================================================================
+%% gen_server callbacks
+%%================================================================================
+
+init({server, Shard, Subscriber}) ->
+    logger:set_process_metadata(#{ domain => [ekka, rlog, bootstrapper, server]
+                                 , shard  => Shard
+                                 }),
+    #{tables := Tables} = ekka_rlog_config:shard_config(Shard),
+    ?tp(info, rlog_bootstrapper_start,
+        #{ shard     => Shard
+         , subscribe => Subscriber
+         }),
+    Queue = replayq:open(#{ mem_only => true
+                          , sizer    => fun(_) -> 1 end
+                          }),
+    self() ! table_loop,
```

Comment:
After some thinking I think in this case it makes sense to use self message here, because the server also does so during normal operation.
Looping over tables/keys is done via self-messages to support some degree of interactivity (e.g. make it possible to use `sys:get_state`)